File: stack/20有效的括号.js
/**
 * @param {string} s
 * @return {boolean}
 */

 var isValid = function(s) {
    const map = new Map([
        [']','['],
        ['}','{'],
        [')','(']
    ])
    
    if(s.length%2 ===1){
        return false
    }
    
    const stack = []
    for(let char of s){
        //当前为右括号 比较当前符号是否和stack末尾匹配
        if(map.has(char)){
            if(!stack.length || stack[stack.length-1] !== map.get(char)){
                return false
            }
            stack.pop()
        }
        //当前左括号 放入stack
        else{
            stack.push(char)
        }
    }

    return !stack.length
};
isValid('()')

File: golang/20.有效的括号.go
package main

func isLeft(s byte)bool{
	if s == '(' || s == '{' || s == '['{
		return true
	}

	return false
}

func isValid(s string) bool {
	stack := []byte{}
	m := map[byte]byte {
		')':'(',
		']':'[',
		'}':'{',
	}
	
	for i:=0; i<len(s); i++{
		if isLeft(s[i]){
			stack = append(stack, s[i])
		}else{
			if len(stack) != 0 && m[s[i]] == stack[len(stack)-1]{
				stack = stack[:len(stack)-1]
			}else{
				return false
			}
		}
	}

	if len(stack) == 0 {
		return true
	}

	return false
}